dev/io.openliberty.microprofile.faulttolerance.telemetry/src/io/openliberty/microprofile/faulttolerance/telemetry/metrics/integration/AbstractMetricRecorderImpl.java

```diff
@@ -0,0 +1,474 @@
+ * <p>
+ * This class initializes all of the metrics for the method up front in the constructor and stores them in fields so that we don't have to register or look up metrics while the
+ * method is running.
+ * <p>
+ * In some cases, where a metric has tags with lots of values, we use an EnumSet to store the metrics for each tag combination.
+ * <p>
+ * In other cases, where a metric has few or no possible tag values, we just use a separate field for each tag combination.
+ */
+public abstract class AbstractMetricRecorderImpl implements MetricRecorder {
+
+    private static final TraceComponent tc = Tr.register(AbstractMetricRecorderImpl.class);
+
+    // Every metric uses this tag to identify the method it's reporting metrics for
+    private final Attributes methodAttribute;
+
+    //Fields for spec Metric
+    private final LongCounter ftInvocationsTotal;
+    private final boolean fallbackDefined;
+
+    //Fields for @Retry Metrics
+    private final LongCounter ftRetryCallsTotal;
+    private final LongCounter ftRetryRetriesTotal;
+
+    //Fields for @Timeout Metrics
+    private final LongCounter ftTimeoutCallsTotal;
+    private final LongHistogram ftTimeoutExecutionDuration;
+
+    //Fields for @CircuitBreaker Metrics
+    private final LongCounter ftCircuitbreakerCallsTotal;
+    @SuppressWarnings("unused") //This object will be passed into open telemetry by the builder and used there.
+    private final ObservableLongCounter ftCircuitbreakerStateTotal;
+    private final LongCounter ftCircuitbreakerOpenedTotal;
+
+    private long openNanos;
+    private long halfOpenNanos;
+    private long closedNanos;
+    private CircuitBreakerState circuitBreakerState = CircuitBreakerState.CLOSED;
+    protected long lastCircuitBreakerTransitionTime;
+
+    private enum CircuitBreakerState {
+        CLOSED,
+        HALF_OPEN,
+        OPEN
+    }
+
+    //Fields for @Bulkhead Metrics
+    private final LongCounter ftBulkheadCallsTotal;
+    @SuppressWarnings("unused") //This object will be passed into open telemetry by the builder and used there.
+    private final ObservableLongGauge ftBulkheadExecutionsRunning;
+    @SuppressWarnings("unused") //This object will be passed into open telemetry by the builder and used there.
+    private final ObservableLongGauge ftBulkheadExecutionsWaiting;
+    private final LongHistogram ftBulkheadRunningDuration;
+    private final LongHistogram ftBulkheadWaitingDuration;
+
+    private LongSupplier queuePopulationSupplier = null;
+    private LongSupplier concurrentExecutionCountSupplier = null;
+
+    public AbstractMetricRecorderImpl(String classAndMethod, String metricPrefix, Meter meter, RetryPolicy retryPolicy, CircuitBreakerPolicy circuitBreakerPolicy,
+                                      TimeoutPolicy timeoutPolicy,
+                                      BulkheadPolicy bulkheadPolicy, FallbackPolicy fallbackPolicy, AsyncType isAsync) {
```

Comment:
We don't need `metricPrefix` here - that was only necessary for Fault Tolerance 2.0 and earlier where we didn't have tags and the method name and class was put on the front of the metric name.